scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ExtensionMethods.scala

```diff
@@ -1,52 +1,63 @@
 import pl.touk.nussknacker.engine.definition.clazz.{ClassDefinitionSet, MethodDefinition}
-import pl.touk.nussknacker.engine.extension.ExtensionMethods.extensionMethodsHandlers
-
-import java.lang.reflect.{Method, Modifier}
-
-class ExtensionsAwareMethodInvoker(classDefinitionSet: ClassDefinitionSet) {
-
-  private val toInvocationTargetConvertersByClass =
-    extensionMethodsHandlers
-      .map(e => e.invocationTargetClass -> e.createConverter(classDefinitionSet))
-      .toMap[Class[_], ToExtensionMethodInvocationTargetConverter[_]]
-
-  def invoke(method: Method)(target: Any, arguments: Array[Object]): Any = {
-    if (toInvocationTargetConvertersByClass.contains(method.getDeclaringClass)) {
-      toInvocationTargetConvertersByClass
-        .get(method.getDeclaringClass)
-        .map(_.toInvocationTarget(target))
-        // Maybe in future we could write some mechanism to invoke extension methods statically. What I mean is to
-        // find correct extension based on target and then implement simple switch to fire method based on name
-        .map(impl => method.invoke(impl, arguments: _*))
-        .getOrElse {
-          throw new IllegalArgumentException(s"Extension method: ${method.getName} is not implemented")
-        }
-    } else {
-      method.invoke(target, arguments: _*)
-    }
+import pl.touk.nussknacker.engine.extension.ExtensionMethods.extensionMethodsDefinitions
+
+import java.util
+import scala.collection.concurrent.TrieMap
+
+class ExtensionMethodResolver(classDefinitionSet: ClassDefinitionSet) extends MethodResolver {
+  private val executorsCache = new TrieMap[(String, Class[_]), Option[MethodExecutor]]()
+
+  override def resolve(
+      context: EvaluationContext,
+      targetObject: Any,
+      methodName: String,
+      argumentTypes: util.List[TypeDescriptor]
+  ): MethodExecutor =
+    maybeResolve(context, targetObject, methodName, argumentTypes).orNull
+
+  def maybeResolve(
+      context: EvaluationContext,
+      targetObject: Any,
+      methodName: String,
+      argumentTypes: util.List[TypeDescriptor]
+  ): Option[MethodExecutor] = {
+    val targetClass = targetObject.getClass
+    executorsCache.getOrElse(
+      (methodName, targetClass), {
+        val maybeExecutor = extensionMethodsDefinitions
+          .filter(_.appliesToClassInRuntime(targetClass))
+          .flatMap(_.createHandler(classDefinitionSet).findMethod(methodName, argumentTypes.size()))
+          .headOption
+          .map(createExecutor)
+        executorsCache.put((methodName, targetClass), maybeExecutor)
+        maybeExecutor
+      }
+    )
   }
 
-}
+  private def createExecutor(method: ExtensionMethod): MethodExecutor = new MethodExecutor {
 
-object ExtensionAwareMethodsDiscovery {
+    override def execute(context: EvaluationContext, target: Any, args: Object*): TypedValue = {
+      new TypedValue(method.invoke(target, args: _*), null)
```

Comment:
Null is fine because it's only used during SpeL conversion and even with null it should work well. By contrast it was easier to write proper implementation than comment.